app/src/processing/app/Editor.java

```diff
@@ -1265,43 +1260,58 @@ public void actionPerformed(ActionEvent e) {
     return menu;
   }
 
+  /**
+   * Wrapper around RSyntaxTextArea's RecordableTextActions. Normally,
+   * these actions use the event source, or the most recently focused
+   * component to find the text area to work on, but this does not
+   * always work. This wrapper makes them always work on the currently
+   * selected tab instead.
+   */
+  class RstaActionWrapper extends ActionWrapper {
+    RstaActionWrapper(RecordableTextAction action) {
+      super(action);
+    }
+
+    @Override
+    public void actionPerformed(ActionEvent e) {
+      ((RecordableTextAction) getWrappedAction()).actionPerformedImpl(e, getCurrentTab().getTextArea());
+    }
+  }
 
   private JMenu buildEditMenu() {
     JMenu menu = new JMenu(tr("Edit"));
     menu.setName("menuEdit");
     menu.setMnemonic(KeyEvent.VK_E);
 
-    undoItem = newJMenuItem(tr("Undo"), 'Z');
+    // Create a dummy RTextArea, to force it to create the static list
+    // of actions accessible through RTextArea.getAction().
+    new RTextArea();
+
+    RecordableTextAction undoAction = RTextArea.getAction(RTextArea.UNDO_ACTION);
+    JMenuItem undoItem = new JMenuItem(new RstaActionWrapper(undoAction));
```

Comment:
The dummy object creation to obtain the `UNDO_ACTION` makes me really uneasy.
Also, we already have a `EditorTab.handleUndo()` method, why not just call that instead of trying to fit the RSTA action?

Moreover the RSTA action is wrapped inside another ActionWrapper class that is made just to route the message to the correct EditorTab through getCurrentTab(), while it may be much more simple to do something like:

```java
undoItem.addActionListener(() -> getCurrentTab().handleUndo());
```

Or maybe I'm missing something?

I remember that I've already solved this in another refactoring experiment I did some time ago I'll try to recover this attempt from my old branches.